cumulus/parachains/runtimes/coretime/coretime-westend/src/coretime.rs

```diff
@@ -16,28 +16,61 @@
 
 use crate::*;
 use codec::{Decode, Encode};
+use core::marker::PhantomData;
 use cumulus_pallet_parachain_system::RelaychainDataProvider;
 use cumulus_primitives_core::relay_chain;
 use frame_support::{
 	parameter_types,
 	traits::{
 		fungible::{Balanced, Credit},
-		OnUnbalanced,
+		Imbalance, OnUnbalanced, TryDrop,
 	},
 };
 use pallet_broker::{
 	CoreAssignment, CoreIndex, CoretimeInterface, OnDemandRevenueRecord, PartsOf57600,
 	RCBlockNumberOf, RevenueInbox,
 };
 use parachains_common::{AccountId, Balance};
+use sp_core::Get;
+use sp_runtime::traits::AccountIdConversion;
 use westend_runtime_constants::system_parachain::coretime;
 use xcm::latest::prelude::*;
 
-pub struct CreditToCollatorPot;
-impl OnUnbalanced<Credit<AccountId, Balances>> for CreditToCollatorPot {
-	fn on_nonzero_unbalanced(credit: Credit<AccountId, Balances>) {
-		let staking_pot = CollatorSelection::account_id();
-		let _ = <Balances as Balanced<_>>::resolve(&staking_pot, credit);
+pub struct BurnAtRelay;
+impl OnUnbalanced<Credit<AccountId, Balances>> for BurnAtRelay {
+	fn on_nonzero_unbalanced(amount: Credit<AccountId, Balances>) {
+		let value = amount.peek();
+
+		// The assets being burnt, from the relay chain perspective
+		let parent_assets: Assets =
+			vec![Asset { id: AssetId(Location::here()), fun: Fungible(value) }].into();
+
+		let send_res = PolkadotXcm::send_xcm(
```

Comment:
Does the `check_out` logic get called somewhere else before this?

Reply:
Not sure I follow. It's the parachain side, so we don't have to check out assets here IIUC. They're checked out at the relay chain, and here, we want the RC to check them back in and burn them immediately.

Reply:
Yeah I understand, but it's up to the implementation of `AssetTransactor`.

Follow the code paths in [do_teleport_assets](https://github.com/paritytech/polkadot-sdk/blob/polkadot-v1.13.0/polkadot/xcm/pallet-xcm/src/lib.rs#L1602). You will see it ends up executing a [teleport_assets_program](https://github.com/paritytech/polkadot-sdk/blob/polkadot-v1.13.0/polkadot/xcm/pallet-xcm/src/lib.rs#L2148-L2168) locally. IMO the system teleporting assets should behave like other teleports. If the `Config::AssetTransactor` changes in a runtime upgrade, it will be strange if these are accounted for differently.

We probably need some refactor to make system teleports more ergonomic with less code duplication, but I would treat these like other teleports for now (except that we can get away with `UnpaidExecution` and extra privilege here). cc @franciscoaguirre @acatangiu

Reply:
yeah I was thinking the same, but the first bump in the road is that both `InitiateTeleport` XCM instructions and/or pallet-xcm/xtokens/etc build XCMs assuming `BuyExecution` and do not "out of the box" support UnpaidExecution.

We should def fix that, but I can't think of a way to do it without breaking pallet APIs (and/or without new XCM version) (as I'm writing this, obviously we can add new unpaid transfer extrinsics - ugly, yes)

I will file an issue

Reply:
Opened https://github.com/paritytech/polkadot-sdk/issues/4854 - lemme know if anyone can pick it up, or I should add it to our priority queue.